### PR TITLE
New notebook from the command line issues

### PR DESCRIPTION
**Pull request by bpasero**
This is for #125028 and not the full fix yet but I want to get some feedback if the direction makes sense. 

The challenge is that we seem to have some logic how to create a untitled notebook URI in [`mainThreadNotebookDocuments.ts`](https://github.com/microsoft/vscode/blob/d0cc52143b35092ebabe55009a79cc8d29e23375/src/vs/workbench/api/browser/mainThreadNotebookDocuments.ts#L141-L149) but we also need it for the untitled editor factory where we have to create a notebook editor:

https://github.com/microsoft/vscode/blob/d3613d0c0cc4952ef7a02002433213d76ef3b81e/src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts#L171-L171

This method is responsible for converting a simple JSON object ("untyped editor input") to a real untitled notebook editor input. With this PR, the behaviour of both are aligned with regards to:
* we ensure an untitled notebook doesn't exist yet by incrementing a counter
* we add the proper suffix via `possibleFileEnding` usage (unless a resource is associated)
* we define a query param (`query: notebookProviderInfo.id`)

One missing aspect is how to forward the information down to where the working copy is being created that we have an untitled resource with associated file path. Basically this code:

https://github.com/microsoft/vscode/blob/ca4bf8a05616559b00f864352619e00d3948d6e6/src/vs/workbench/contrib/notebook/common/notebookEditorModel.ts#L519-L519

Needs to change in a way that it continues to call `resolve({ untitledResource: this.resource })` for normal untitled editors and `resolve({ associatedResource: this.resource })` for untitled with associated path. This distinction makes it possible to get the right behaviour which is:
* untitled with associated path do not ask for a file path when saving
* they are dirty from the beginning

One idea I had was to also encode this info into the URI as query param, but I am not sure where this resource is travelling and if that is a safe thing to do.

Thoughts?

### CODE REVIEW

File: src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts [hunk 1]
```diff
@@ -48,12 +48,11 @@ export class NotebookProviderInfoStore extends Disposable {
 	private readonly _memento: Memento;
 	private _handled: boolean = false;
 
-	private _untitledCounter = 1;
-
 	private readonly _contributedEditors = new Map<string, NotebookProviderInfo>();
 	private readonly _contributedEditorDisposables = new DisposableStore();
 
 	constructor(
+		@INotebookService private readonly _notebookService: INotebookService,
```
Comment (jrieken) [inline src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts:55]: This is a cyclic dependency between the notebook service and this type. How about some indirection like a callback, e.g `existsCallback: (uri: URI) => boolean`?

**Comment by jrieken** [comment]
> Thoughts?

Idk - It's the first time I hear about `UntitledEditorInputFactoryFunction`. Is the contract that it must be created with the final untitled URI or can the URI change later? In a way, I am wishing for generating the untitled-uri right before creating the model, like [here](https://github.com/microsoft/vscode/blob/9bb711d3b6a56de77fa41e6b556d2e24ebe50463/src/vs/workbench/contrib/notebook/common/notebookEditorModelResolverServiceImpl.ts#L155). And only have the logic once, e.g not in mainThreadNotebook and the factory function. 

I think this needs further clarification. Let's reserve some 1-1 time tomorrow or next week

**Comment by bpasero** [comment]
@jrieken the factory was added only recently (last week?) to distinguish normal editors from untitled editors (and diff editors have their own factory too), but it doesn't really matter and is just there for convenience.

The underlying contract is that you get a lightweight "untyped" JSON editor input and have to return a heavy "typed" editor input that is then opened in an editor. There are only a few untyped editors:
* the 99% case: just a resource (`{ resource: ...}`)
* the untitled case: most often `{}`, in some cases a resource is provided (only when e.g. run from command line with a path)
* the diff editor case: looks mostly like this: `{ originalInput: URI, modifiedInput: URI }`

The URI of the editor input does not have to be known up front, in fact it is even optional for editor inputs so notebooks are free to do whatever they want, even if that means to resolve the URI only in the `resolve` method of the input. 

1on1 sounds good, we can do next week since this probably is not making it for June anyway.

**Comment by bpasero** [comment]
Btw I like the idea of having this logic somewhere here:

https://github.com/microsoft/vscode/blob/9bb711d3b6a56de77fa41e6b556d2e24ebe50463/src/vs/workbench/contrib/notebook/common/notebookEditorModelResolverServiceImpl.ts#L155

It would however require that for the `aquire` here:

https://github.com/microsoft/vscode/blob/9bb711d3b6a56de77fa41e6b556d2e24ebe50463/src/vs/workbench/contrib/notebook/common/notebookEditorModelResolverServiceImpl.ts#L183

We can pass on additional context to pipe through the untitled with associated file path case. Could be just a boolean too.

**Comment by bpasero** [comment]
Continues in https://github.com/microsoft/vscode/pull/130236, I can walk you through when you are back.

**Pull request closed without merging**